File: Data Structures - Algorithms/Code_DataStructure/DataStructureImplementation/Stack/FromArray/stack.cs
using System;

namespace ImplementStack
{
class Stack
{
	private int[] arr;
	private int top;			// Index of the top item
	private int max_size;

	public Stack(int size)
	{
		arr = new int[size]; 	// Maximum size of Stack
		top = -1;
		max_size = size;
	}

	public void push(int item)
	{
		if (top == max_size - 1) {
			Console.WriteLine("Stack Overflow");
			return;
		}
		else {
			arr[++top] = item;
		}
	}

	public int pop()
	{
		if (top == -1) {
			Console.WriteLine("Stack is Empty");
			return -1;
		}
		else {
			Console.WriteLine("{0} popped from stack ", arr[top]);
			return arr[top--];
		}
	}

	public int peek()
	{
		if (top == -1) {
			Console.WriteLine("Stack is Empty");
			return -1;
		}
		else {
			Console.WriteLine("{0} popped from stack ", arr[top]);
			return arr[top];
		}
	}

	public void printStack()
	{
		if (top == -1) {
			Console.WriteLine("Stack is Empty");
			return;
		}
		else {
			for (int i = 0; i <= top; i++) {
				Console.WriteLine("{0} pushed into stack", arr[i]);
			}
		}
	}
}

// Driver program to test above functions
class Program
{
	static void Main()
	{
		Stack p = new Stack(5);

		p.push(10);
		p.push(20);
		p.push(30);
		p.printStack();
		p.pop();
	}
}
}

File: Data Structures - Algorithms/Demo/DataStructures/Stack/ImplStackFromArray/stack.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAX_SIZE 1000

class Stack
{
private:
	int top;					// Index of the top item

public:
	int arr[MAX_SIZE];			// Maximum size of Stack

	Stack() {
		top = -1;
	}

	bool push(int x) {
		if (isFull()) {
			cout << "Stack Overflow";
			return false;
		}
		else {
			arr[++top] = x;
			cout << x << " pushed into stack\n";
			return true;
		}
	}

	int pop() {
		if (top < 0) {
			cout << "Stack Underflow";
			return 0;
		}
		else {
			int x = arr[top--];
			return x;
		}
	}

	int peek() {
		if (isEmpty()) {
			cout << "Stack is Empty";
			return 0;
		}
		else {
			int x = arr[top];
			return x;
		}
	}

	bool isEmpty() { return (top < 0); }

	bool isFull() { return (top >= (MAX_SIZE - 1)); }
};

int main()
{
	class Stack s;
	s.push(10);
	s.push(20);
	s.push(30);
	
	// print top element of stack after popping
	cout << "Top element is: " << s.peek() << endl;
	
	// print all elements in stack:
	cout << "Elements present in stack: ";
	while(!s.isEmpty()) {
		cout << s.peek() <<" ";
		s.pop();
	}

	return 0;
}
